src/CodeGen_ARM.cpp

```diff
@@ -1261,4 +1261,14 @@ int CodeGen_ARM::native_vector_bits() const {
     return 128;
 }
 
+bool CodeGen_ARM::target_needs_software_cast_from_float16_to(Type t) const {
```

Comment:
> Are we ever going to encounter a target with hardware support for casts from float16 but not hardware support for casts to float16?

Yes, sort of... There are two methods because of 

- how this PR handles the conversion behavior on x86. It is **not symmetrical**.
-  When downcasting **we need to know the rounding mode**, but it's irrelevant we upcasting.

Take a look at the contents of ``CodeGen_X86::target_needs_software_cast_from_float16_to(Type t)`` and ``CodeGen_X86::target_needs_software_cast_to_float16_from(Type t, RoundingMode rm)`` to understand this.

Briefly though, the ``double -> float16`` and ``float16 -> double`` conversions are **not** symmetrical in my implementation on x86.

* ``float16 -> double`` we can do in hardware because we can use ``vcvtph2ps`` (x86-64 fp16c intrinsic) to get a single precision and then convert (``fpext`` LLVM IR instruction) to double. Upcasting twice is legal because there is no loss in precision.

* ``double -> float16`` cannot be done in hardware because ``vcvtph2ps`` only works with ``float`` and rounding twice (``double`` -> ``float`` -> ``float16``) is not acceptable because that would not respect the rounding mode. Therefore the software implementation is **always used here**.

* ``float -> float16`` can be done in hardware provided the rounding mode is not "round to nearest, ties to away". All other rounding modes are supported by ``vcvtps2ph`` but the other rounding mode must be handled by the software implementation.

* ``float16 -> float`` can be done in hardware using the ``vcvtps2ph`` intrinsic.